### PR TITLE
Robust handling of PFS cookies

### PR DESCRIPTION
**Pull request by nl0**
## Description

Lazily (re)set PFS cookie when required

## TODO

- [x] [Changelog](../tree/master/docs/CHANGELOG.md) entry (skip if change is not significant to end users, e.g. docs only)

### CODE REVIEW

File: catalog/app/utils/PFSCookieManager.tsx [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  if (state.expires.getTime() + params.histeresis < params.expires.getTime()) return true
+  return false
+}
+
+export function PFSCookieManager({ children }: React.PropsWithChildren<{}>) {
+  const stateRef = React.useRef<State>()
+
+  const req = useApi()
+
+  const setPFSCookie = React.useCallback(
```
Comment (fiskus) [inline catalog/app/utils/PFSCookieManager.tsx:45]: Maybe, `requestPFSCookie`, so not to confuse with `React.setState` and highlight that this function is not cheap

File: catalog/app/utils/PFSCookieManager.tsx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+const selectToken = createSelector(
+  tokensSelector,
+  (tokens) => tokens?.token as string | undefined,
+)
+
+interface State {
+  token: string | undefined
+  expires: Date
+  promise: Promise<void>
```
Comment (fiskus) [inline catalog/app/utils/PFSCookieManager.tsx:23]: Would be nice to have some clue what is it for. For example,

```ts
State {
 ...
 settingCookie: Promise<void>
}
```

**Review by fiskus — COMMENTED** [review]
For some reason, it still doesn't work for me
Screenshot from 2024-04-25 15-53-52

**Comment by nl0** [comment]
> For some reason, it still doesn't work for me

1. you need an updated registry for that to work (PR in review)
2. it won't work on localhost bc the cookie has strict policy -- it only works for the same (sub?)domain

**Review by fiskus — DISMISSED** [review]

File: catalog/app/utils/PFSCookieManager.tsx (hunk 1 above)
Comment (fiskus) [reply catalog/app/utils/PFSCookieManager.tsx:32]: I think, you can make this callback an ordinary function (more likely two functions), but probably it's not worth it, and not much time left

**Review by fiskus — APPROVED** [review]

File: catalog/app/utils/PFSCookieManager.tsx (hunk 1 above)
Comment (nl0) [reply catalog/app/utils/PFSCookieManager.tsx:32]: if i split them, nothing except the actual request call will remain in the "ordinary" function

**Review by fiskus — APPROVED** [review]

**Pull request merged**